### PR TITLE
feat: add support for HunYuanDit ControlNet

### PR DESCRIPTION
**Pull request by CrazyBoyM**
this pr add support for HunYuanDit ControlNet.

### Test Example:
image
image
image
image

### Here is the official exmaple:
image

### You can download controlnet model weight files here:
https://huggingface.co/Tencent-Hunyuan/HYDiT-ControlNet-v1.2/tree/main
image

### Here is the workflow file:
[hunyuan-controlnet-workflow.json](https://github.com/user-attachments/files/16513972/hunyuan-controlnet-workflow.json)

### CODE REVIEW

**Comment by CrazyBoyM** [comment]
full workflow:
image

**Comment by CrazyBoyM** [comment]
Base model:   
https://huggingface.co/comfyanonymous/hunyuan_dit_comfyui/blob/main/hunyuan_dit_1.2.safetensors

**Comment by CrazyBoyM** [comment]
@comfyanonymous could you review this PR?

File: comfy/latent_formats.py [hunk 1]
```diff
@@ -37,6 +37,20 @@ def __init__(self):
                 ]
         self.taesd_decoder_name = "taesdxl_decoder"
 
+class HunyuanDit(LatentFormat):
```
Comment (mcmonkey4eva) [inline comfy/latent_formats.py]: Why is the separate latent format needed? This is just a copypaste of sdxl

File: comfy/ldm/hydit/controlnet.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+        # ========================= Forward pass through HunYuanDiT blocks =========================
+        controls = []
+        x = x + self.before_proj(condition) # add condition
+        for layer, block in enumerate(self.blocks):
+            x = block(x, c, text_states, freqs_cis_img)
+            controls.append(self.after_proj_list[layer](x)) # zero linear for output
+
+
+        return {'output': controls}
+
```
Comment (mcmonkey4eva) [inline comfy/ldm/hydit/controlnet.py]: file format's wonky at the bottom here, check for spacing errors

File: comfy/latent_formats.py (hunk 1 above)
Comment (CrazyBoyM) [reply comfy/latent_formats.py]: I had removed it.

File: comfy/ldm/hydit/controlnet.py (hunk 2 above)
Comment (CrazyBoyM) [reply comfy/ldm/hydit/controlnet.py]: fix in right format.

**Comment by CrazyBoyM** [comment]
condition image for test:
canny
asset_input_pose
asset_input_depth

**Comment by CrazyBoyM** [comment]
can&nbsp;you&nbsp;review&nbsp;this&nbsp;pr&nbsp;again？
来新璐



---Original---
From: ***@***.***&gt;
Date: Wed, Aug 7, 2024 02:33 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [comfyanonymous/ComfyUI] feat: add support for HunYuanDitControlNet (PR #4245)




 
@mcmonkey4eva commented on this pull request.
 
 
In comfy/latent_formats.py:
 &gt; @@ -37,6 +37,20 @@ def __init__(self):                  ]          self.taesd_decoder_name = &quot;taesdxl_decoder&quot;   +class HunyuanDit(LatentFormat):  
Why is the separate latent format needed? This is just a copypaste of sdxl
 
&mdash;

File: comfy/ldm/hydit/controlnet.py (hunk 2 above)
Comment (mcmonkey4eva) [reply comfy/ldm/hydit/controlnet.py]: it's still wrong

Comment (CrazyBoyM) [reply comfy/ldm/hydit/controlnet.py]: do we have a `pre-commit` tool?

Comment (CrazyBoyM) [reply comfy/ldm/hydit/controlnet.py]: I had used 'black' to format it.

**Comment by CrazyBoyM** [comment]
do we have a 'pre-commit' for auto code style format?

**Comment by comfyanonymous** [comment]
I ended up changing this a bit, the node to use this controlnet is now the: "ControlNetApply SD3 and HunyuanDiT" node.

https://github.com/comfyanonymous/ComfyUI/commit/a475ec2300abb4eab845510ad0da596114174274

**Comment by CrazyBoyM** [comment]
> I ended up changing this a bit, the node to use this controlnet is now the: "ControlNetApply SD3 and HunyuanDiT" node.
> 
> [a475ec2](https://github.com/comfyanonymous/ComfyUI/commit/a475ec2300abb4eab845510ad0da596114174274)

it sames that the strength_type is not able to work? i see it is defined in the Controlnet class but never been used

**Comment by CrazyBoyM** [comment]
> > I ended up changing this a bit, the node to use this controlnet is now the: "ControlNetApply SD3 and HunyuanDiT" node.
> > [a475ec2](https://github.com/comfyanonymous/ComfyUI/commit/a475ec2300abb4eab845510ad0da596114174274)
> 
> it sames that the strength_type is not able to work? i see it is defined in the Controlnet class but never been used

oh, sorry, i see it is used in the control_merge function with "x *= (self.strength ** float(len(control_output) - i))", but in face, i keep both the `strength_weight` and `strength` is aim to let users can use both method for scale control signals, my friend tell me he can not use the original `stength` now. can we add a switch option for `ControlNetApplySD3`?

**Comment by anup-kaushal** [comment]
Not working Comfyui new update has no node named

**Pull request merged**